// @morpheus/core/dist/index.js
class EventEmitter {
    events = {};
    on(event, listener) {
        if (typeof this.events[event] !== "object")
            this.events[event] = [];
        this.events[event].push(listener);
        return () => this.off(event, listener);
    }
    off(event, listener) {
        if (typeof this.events[event] !== "object")
            return;
        const idx = this.events[event].indexOf(listener);
        if (idx > -1)
            this.events[event].splice(idx, 1);
    }
    removeAllListeners() {
        Object.keys(this.events).forEach((event) => this.events[event].splice(0, this.events[event].length));
    }
    emit(event, ...args) {
        if (typeof this.events[event] !== "object")
            return;
        this.events[event].forEach((listener) => listener.apply(this, args));
    }
    once(event, listener) {
        const remove = this.on(event, (...args) => {
            remove();
            listener.apply(this, args);
        });
    }
}
class LifecycleEvent extends EventEmitter {
    _initialize = false;
    init() {
        const factory = (key) => {
            return (e) => {
                this.emit(key, e);
            };
        };
        M.onReady(factory("ready"))
            .onDestroy(factory("destroy"))
            .onHide(factory("hide"))
            .onBack(factory("back"))
            .onRestore(factory("restore"))
            .onPause(factory("pause"))
            .onResume(factory("resume"));
        this._initialize = true;
    }
    on(event, listener) {
        if (this._initialize === false)
            this.init();
        return super.on(event, listener);
    }
    off(event, listener) {
        return super.off(event, listener);
    }
    emit(event, ...args) {
        super.emit(event, ...args);
    }
    once(event, listener) {
        super.once(event, listener);
    }
}
const lifecycle = new LifecycleEvent();

// lifecycle export with EventEmitter
/**
 * Page API
 */
function getPageInfo() {
    return M.page.info();
}
function setPageDefer(enabled, time) {
    M.page.defer(enabled, time);
}
function movePage(options) {
    M.page.html(options);
}
function backPage(options) {
    M.page.back(options);
}
function removePage(url) {
    M.page.remove(url);
}
function replacePage(options) {
    M.page.replace(options);
}
function moveActivity(options) {
    M.page.activity(options);
}
/**
 * Security API
 */
function encrypt(source) {
    return M.sec.encrypt(source);
}
function decrypt(source) {
    return M.sec.decrypt(source);
}
/**
 * Data API
 */
function getParamData(key) {
    return M.data.param(key);
}
function setParamData(key, value) {
    M.data.param(key, value);
}
function removeParamData(key) {
    M.data.removeParam(key);
}
function removeAllParamData() {
    M.data.removeParam();
}
function getGlobalData(key) {
    return M.data.global(key);
}
function setGlobalData(key, value) {
    M.data.global(key, value);
}
function removeGlobalData(key) {
    M.data.removeGlobal(key);
}
function removeAllGlobalData() {
    M.data.removeGlobal();
}
function setStorageData(key, value) {
    M.data.storage(key, value);
}
function getStorageData(key) {
    return M.data.storage(key);
}
function removeStorageData(key) {
    M.data.removeStorage(key);
}
function removeAllStorageData() {
    M.data.removeStorage();
}
/**
 * Navigator API
 */
function getOs() {
    return M.navigator.os();
}
function getDevice() {
    return M.navigator.device();
}
function getBrowser() {
    return M.navigator.browser();
}
/**
 * Info API
 */
function getLibrariesVersion() {
    return M.info.version();
}
function getMemoryInfo() {
    return M.info.memory();
}
function getDeviceDetail() {
    return M.info.device();
}
function getStackInfo() {
    return M.info.stack();
}
function getAppInfo() {
    return M.info.app();
}
/**
 * Apps API
 */
function openBrowser(url, encoding) {
    return M.apps.browser(url, encoding);
}
function uninstallApp(identifier) {
    M.apps.remove(identifier);
}
function installApp(apkUrl, appName, options) {
    M.apps.downloadAndInstall(apkUrl, appName, options);
}
function getInstalledAppInfo(identifier) {
    return M.apps.info(identifier);
}
function openApp(scheme, param) {
    M.apps.open(scheme, param);
}
function openStoreApp(scheme) {
    M.apps.store(scheme);
}
/**
 * System API
 */
function openSystemCall(phoneNumber) {
    M.sys.call(phoneNumber);
}
function exitApp() {
    M.sys.exit();
}
function setSystemFlash(status) {
    M.sys.flash(status);
}
function getSystemFlash() {
    return M.sys.flash();
}
function openSystemMail(options) {
    M.sys.mail(options);
}
function openSystemSms(options) {
    M.sys.sms(options);
}
function vibration(time) {
    M.sys.vibration(time);
}
/**
 * Tool API (logger)
 */
function nativeLog(options) {
    M.tool.log(options);
}
/**
 * Native API Executor
 */
function execute(apiKey, ...params) {
    return M.execute(apiKey, ...params);
}

export { EventEmitter, backPage, decrypt, encrypt, execute, exitApp, getAppInfo, getBrowser, getDevice, getDeviceDetail, getGlobalData, getInstalledAppInfo, getLibrariesVersion, getMemoryInfo, getOs, getPageInfo, getParamData, getStackInfo, getStorageData, getSystemFlash, installApp, lifecycle, moveActivity, movePage, nativeLog, openApp, openBrowser, openStoreApp, openSystemCall, openSystemMail, openSystemSms, removeAllGlobalData, removeAllParamData, removeAllStorageData, removeGlobalData, removePage, removeParamData, removeStorageData, replacePage, setGlobalData, setPageDefer, setParamData, setStorageData, setSystemFlash, uninstallApp, vibration };
