// @morpheus/cli/dist/AppConfig.BjmXUQ7C.js
import * as path from 'path';
import { l as libExports } from './vendor.DEZAgNuY.js';
import { a as BaseJson, P as PackageJson, f as fatal } from './index.BBihg1b8.js';

class AppExtendJson extends BaseJson {
}

class AppConfig {
    constructor(rootDir) {
        this.rootDir = rootDir;
        this._extendConfig = new AppExtendJson(path.join(rootDir, 'morpheus.config.json'));
        this._packageJson = new PackageJson(path.join(rootDir, 'package.json'));
    }
    get config() {
        return this._extendConfig;
    }
    get packageJson() {
        return this._packageJson;
    }
    get extendConfig() {
        return this._extendConfig.json;
    }
    get androidDir() {
        return this.extendConfig.androidRootDir ? path.join(this.rootDir, this.extendConfig.androidRootDir) : this.defaultAndroidDir();
    }
    get iosDir() {
        return this.extendConfig.iosRootDir ? path.join(this.rootDir, this.extendConfig.iosRootDir) : this.defaultIosDir();
    }
    get assetDir() {
        return this.defaultAssetsDir();
    }
    get webDir() {
        return this.extendConfig.webRootDir ? path.join(this.rootDir, this.extendConfig.webRootDir) : this.rootDir;
    }
    get legacyJsOutPut() {
        return this.extendConfig.legacyJsOutPut ? path.join(this.rootDir, this.extendConfig.legacyJsOutPut) : path.join(this.assetDir, 'res', 'www', 'js');
    }
    get manifestExt() {
        return this.extendConfig.manifestExt || 'mcore.mobile.app.implementation.ExtendWNInterface';
    }
    get appLibrariesDir() {
        return path.join(this.rootDir, '.morpheus', 'libs');
    }
    defaultAndroidDir() {
        return path.join(this.rootDir, 'android', 'app');
    }
    defaultIosDir() {
        return path.join(this.rootDir, 'ios', 'project');
    }
    defaultAssetsDir() {
        return path.join(this.rootDir, 'assets');
    }
    async load() {
        if (!(await libExports.pathExists(this._extendConfig.path))) {
            fatal(`${this._extendConfig.path} is not exist.`);
        }
        if (!(await libExports.pathExists(this._packageJson.path))) {
            fatal(`${this._packageJson.path} is not exist.`);
        }
        await this._extendConfig.read();
        await this._packageJson.read();
    }
}

export { AppConfig as A };
