// @morpheus/addon-db/dist/index.js
function create(path) {
    return new Promise((resolve) => {
        M.db.create({
            path: path,
            callback: (status, result) => {
                resolve(result);
            },
        });
    });
}
function open(path) {
    return new Promise((resolve) => {
        M.db.open({
            path: path,
            callback: (status, result) => {
                resolve(result);
            },
        });
    });
}
function close(path) {
    return new Promise((resolve) => {
        M.db.close({
            path: path,
            callback: (status, result) => {
                resolve(result);
            },
        });
    });
}
function remove(path) {
    return new Promise((resolve) => {
        M.db.remove({
            path: path,
            callback: (status, result) => {
                resolve(result);
            },
        });
    });
}
function execute(options) {
    return new Promise((resolve) => {
        M.db.execute({
            ...options,
            callback: (status, result) => {
                resolve(result);
            },
        });
    });
}

export { close, create, execute, open, remove };
