// @morpheus/addon-net/dist/index.js
function clearCookie() {
    M.net.cookie.clear();
}
function sendHttp(options) {
    return new Promise((resolve, reject) => {
        M.net.http.send({
            ...options,
            success: (receivedData) => {
                resolve(receivedData);
            },
            error: (code, message) => {
                reject(new Error(`[${code}] ${message}`));
            },
        });
    });
}
function sendSocket(options) {
    return new Promise((resolve, reject) => {
        M.net.socket.send({
            ...options,
            success: (receivedData) => {
                resolve(receivedData);
            },
            error: (code, message) => {
                reject(new Error(`[${code}] ${message}`));
            },
        });
    });
}
function updateResource(options) {
    return new Promise((resolve) => {
        M.net.res.update({
            ...options,
            progress: (...args) => {
                if (options && options.progress) {
                    options.progress(...args);
                }
            },
            finish: (status, info) => {
                resolve({ status, info });
            },
        });
    });
}
function checkResource(options) {
    return new Promise((resolve) => {
        M.net.res.check({
            ...options,
            callback: (status, info) => {
                resolve({ status, info });
            },
        });
    });
}
function resetResourceVersion() {
    M.net.res.reset();
}

export { checkResource, clearCookie, resetResourceVersion, sendHttp, sendSocket, updateResource };
