// @morpheus/addon-media/dist/index.js
function camera(options) {
    return new Promise((resolve) => {
        M.media.camera({
            ...options,
            callback: (_, result) => {
                resolve(result);
            },
        });
    });
}
function album(path) {
    return new Promise((resolve) => {
        M.media.album({
            path,
            callback: (_, result) => {
                resolve(result);
            },
        });
    });
}
function picker(options) {
    return new Promise((resolve) => {
        M.media.picker({
            ...options,
            callback: (_, result) => {
                resolve(result);
            },
        });
    });
}
function removeLibrary(options) {
    return new Promise((resolve) => {
        M.media.removeLibrary({
            ...options,
            callback: (_, result) => {
                resolve(result);
            },
        });
    });
}
function play(options) {
    M.media.play(options);
}
function record(options) {
    return new Promise((resolve) => {
        M.media.record({
            ...options,
            callback: (_, result) => {
                resolve(result);
            },
        });
    });
}
function getMediaByPath(path) {
    return M.media.get(path);
}
function optimize(options) {
    return new Promise((resolve) => {
        M.media.optimize({
            ...options,
            callback: (result) => {
                resolve(result);
            },
        });
    });
}
function crop(options) {
    return new Promise((resolve) => {
        M.media.crop({
            ...options,
            callback: (result) => {
                resolve(result);
            },
        });
    });
}

export { album, camera, crop, getMediaByPath, optimize, picker, play, record, removeLibrary };
