// @morpheus/addon-file/dist/index.js
function create(options) {
    return M.file.create(options);
}
function read(options) {
    return new Promise((resolve) => {
        M.file.read({
            ...options,
            callback: (status, result) => {
                resolve(result);
            },
        });
    });
}
function remove(path) {
    return M.file.remove(path);
}
function info(path) {
    return M.file.info(path);
}
function list(path, filter) {
    return M.file.list(path, filter);
}
function copy(options) {
    return new Promise((resolve) => {
        M.file.copy({
            ...options,
            progress: (total, current, index, count, srcPath, dstPath) => {
                if (typeof options.progress === "function") {
                    options.progress({ total, current, index, count, srcPath, dstPath });
                }
            },
            finish: (status, result) => {
                resolve(result);
            },
        });
    });
}
function move(options) {
    return new Promise((resolve) => {
        M.file.move({
            ...options,
            progress: (total, current, index, count, srcPath, dstPath) => {
                if (typeof options.progress === "function") {
                    options.progress({ total, current, index, count, srcPath, dstPath });
                }
            },
            finish: (status, result) => {
                resolve(result);
            },
        });
    });
}
function write(options) {
    return new Promise((resolve) => {
        M.file.write({
            ...options,
            callback: (status, result) => {
                resolve(result);
            },
        });
    });
}
function picker(options) {
    return new Promise((resolve) => {
        M.file.picker({
            ...options,
            callback: (status, result) => {
                resolve(result);
            },
        });
    });
}

export { copy, create, info, list, move, picker, read, remove, write };
